x/evm/wrappers/bank.go

```diff
@@ -0,0 +1,104 @@
+// Copyright Tharsis Labs Ltd.(Evmos)
+// SPDX-License-Identifier:ENCL-1.0(https://github.com/evmos/evmos/blob/main/LICENSE)
+
+package wrappers
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/evmos/evmos/v20/x/evm/config"
+	"github.com/evmos/evmos/v20/x/evm/types"
+)
+
+var _ types.BankWrapper = BankWrapper{}
+
+// BankWrapper is a wrapper around the Cosmos SDK bank keeper
+// that is used to manage an evm denom a custom decimal representation.
+// The wrapper makes the corresponding conversions to achieve:
+//   - With the EVM, the wrapper works always with 18 decimals.
+//   - With the Cosmos bank module, the wrapper works always
+//     with the bank module decimals.
+type BankWrapper struct {
+	types.BankKeeper
+}
+
+// NewBankWrapper creates a new BankWrapper instance.
+func NewBankWrapper(
+	bk types.BankKeeper,
+) *BankWrapper {
+	return &BankWrapper{
+		bk,
+	}
+}
+
+// GetBalance returns the balance of the given account converted to 18 decimals.
+// TODO: why do we allow to pass a denom if we can, and want, to handle only the
+// evm denom?
+func (w BankWrapper) GetEVMCoinBalance(ctx context.Context, addr sdk.AccAddress) (sdk.Coin, error) {
```

Comment:
Before removing the `GetBalance` from the wrapper we need to make sure it is not used anywhere within the codebase. Personally I would leave it to be consistent with the purpose of the wrapper. Otherwise we'll get incorrect balance with the evm denom with 6 dec

Reply:
> // TODO: why do we allow to pass a denom if we can, and want, to handle only the
> // evm denom?

If we comply with the bank keeper interface, we can use this bank wrapper across other sections of the codebase, e.g. anteHandlers, postHandlers

Reply:
Good point @GAtom22, will revert this change!

Reply:
Good point @GAtom22, I agree with the purpose of the wrapper. However, since we are performing operation with Mul and Quo we could have over/underflow, and since GetBalance cannot return an error, it was the only approach that came to my mind. Do you have another approach to suggest?

Reply:
I mean, another way to handle the over/underflow gracefully without panic

Reply:
the division `SafeQuo` will return an error if there's a division by zero. Will never occur becuase the conversion factor is always > 0. There cannot be an underflow. For the overflow case, we can handle it by capping the maximum amount. Whenever we have an overflow, we just set the amount to the max number possible. This should never happen ideally, but we can handle this edge case like this

Reply:
I agree, is it still better to have a panic in an extreme case without causing transfer of wrong amounts than risk to use the wrong GetBalance function but recovering all panics! 👍